Add aws-k8s-1.22, vmware-k8s-1.22, aws-k8s-1.22-nvidia, metal-k8s-1.22 variants

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Resolves https://github.com/bottlerocket-os/bottlerocket/issues/1961

**Description of changes:**
For reference, the addition of k8s-1.21 variants was in https://github.com/bottlerocket-os/bottlerocket/pull/1612

```

    vmware-k8s-1.22: use unified layout for vmware-k8s-1.22
    
    Having two disks in the OVF template can trigger sporadic issues when
    importing into the vSphere content library. Switching to the unified
    layout should avoid this failure mode.
    
    Signed-off-by: Ben Cressey <bcressey@amazon.com>


```

```

    Add metal-k8s-1.22 variant
    
    This change adds an additional variant `metal-k8s-1.22`, which includes
    necessary Kubernetes packages and settings for running Bottlerocket on
    metal in a Kubernetes v1.22 cluster.

```

```
    Add aws-k8s-1.22-nvidia variant
    
    Adds aws-k8s-1.22-nvidia variant, create symlinks for previous k8s
    version nvidia variant in models.

```
```
    Add vmware-k8s-1.22 variant
    
    Adds vmware-k8s-1.22 variant, relinks symlinks in models

```
```
    Add aws-k8s-1.22 variant
    
    Adds aws-k8s-1.22 variant, relinks symlinks in models

```
```
    packages: add kubernetes-1.22
    
    Use K8s 1.22 source from EKS-D

```

**Testing done:**

- [x] Pods run OK
- [x]  Conformance testing with 1.22 control plane and x86_64 aws-k8s-1.22 nodes
```
      "progress": {
        "name": "e2e",
        "node": "global",
        "timestamp": "2022-02-23T02:59:32.923164651Z",
        "msg": "Test Suite completed",
        "total": 346,
        "completed": 346
      }
```
- [x]  Conformance testing with 1.22 control plane and aarch64 aws-k8s-1.22 nodes
- [x]  Conformance testing with 1.22 control plane and x86_64 aws-k8s-1.22 nodes in an IPv6 cluster
- [x]  Conformance testing with 1.22 control plane and x86_64 vmware-k8s-1.22 nodes
- [x] Confirm the EBS CSI driver still works.
- [x] Confirm the vSphere CSI driver still works
- [x] Smoke test aws-k8s-1.22-nvidia variant with some GPU workloads
- [x] Metal K8s 1.22 testing
- [x] Check kubelet logs and system logs in general, compared them with the aws-k8s-1.21 variant kubelet logs:


... Others TBD

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
packages: add kubernetes-1.22

Use K8s 1.22 source from EKS-D

etungsten committed:
Add aws-k8s-1.22 variant

Adds aws-k8s-1.22 variant, relinks symlinks in models

etungsten committed:
Add vmware-k8s-1.22 variant

Adds vmware-k8s-1.22 variant, relinks symlinks in models
Update example vmware k8s version in quickstart to v1.21

cbgbt commented:
Early nit: looks like the commit message that adds the 1.22 package incorrectly lists the version as 1.21.

cbgbt reviewed (approved):
This looks good to me, pending testing.

etungsten committed:
Add aws-k8s-1.22-nvidia variant

Adds aws-k8s-1.22-nvidia variant, create symlinks for previous k8s
version nvidia variant in models.

etungsten commented on .github/workflows/build.yml:
```diff
@@ -61,6 +65,22 @@ jobs:
             arch: aarch64
             supported: true
             fetch-upstream: "true"
+          - variant: aws-k8s-1.22
+            arch: x86_64
+            supported: true
+            fetch-upstream: "true"
+          - variant: aws-k8s-1.22
+            arch: aarch64
+            supported: true
+            fetch-upstream: "true"
```
These are temporary fetching from upstream source until EKS-D K8s 1.22 source tarball is finalized.

etungsten committed:
Add metal-k8s-1.22 variant

This change adds an additional variant `metal-k8s-1.22`, which includes
necessary Kubernetes packages and settings for running Bottlerocket on
metal in a Kubernetes v1.22 cluster.

bcressey commented:
nit: commit for `metal-k8s-1.22` says 1.21 in the subject

bcressey commented on QUICKSTART-VMWARE.md:
````diff
@@ -46,8 +46,8 @@ sha512sum -c <<<"e9b1ea5f9b4f95c9b55edada4238bf00b12845aa98bdd2d3edb63ff82a03ada
 ```
-VERSION="v1.2.0"
-VARIANT="vmware-k8s-1.21"
+VERSION="v1.6.0"
````
nit: 1.6.2 or 1.7.0?

etungsten replied:
I changed the example version to point to `vmware-k8s-v1.21` instead. We probably want that anyways.

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -21,23 +21,39 @@ The `#[model]` attribute on Settings and its sub-structs reduces duplication and
+* [Model](src/aws-k8s-1.22-nvidia/mod.rs)
+* [Default settings](src/aws-k8s-1.22-nvidia/defaults.d/)
+
+
```
nit: extra newline

bcressey reviewed (approved):
Nice work! I know there's a staggering amount of copy/paste and fiddly edits but it all looks right to me.

etungsten commented:
Push above rebases onto develop to pull in latest changes.

etungsten commented:
Push above addresses @bcressey 's comments.

zmrow reviewed (approved):
:label:

etungsten commented:
Push above updates the k8s-1.22 package to the latest EKS-D dev release.

etungsten commented:
Push above rebases onto develop and fixes conflicts

etungsten commented:
Push above flips the `kubernetes-1.22` package to use the EKS-D prod release URL for its sources.

Commit:
vmware-k8s-1.22: use unified layout for vmware-k8s-1.22

Having two disks in the OVF template can trigger sporadic issues when
importing into the vSphere content library. Switching to the unified
layout should avoid this failure mode.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

etungsten commented:
Push above adds a commit for changing the image layout for vmware-k8s-1.22 to unified.

Tested the vmware-k8s-1.22 variant and conformance tests passed.

bcressey commented on QUICKSTART-VMWARE.md:
```diff
@@ -1,6 +1,6 @@
 # Using Bottlerocket as a Kubernetes worker node with VMware
 
-> The [VMware variants](variants/README.md#vmware-k8s-121-vmware-kubernetes-121-node) are currently in preview phase and we're looking for [feedback](https://github.com/bottlerocket-os/bottlerocket#contact-us).
+> The [VMware variants](variants/README.md#vmware-k8s-122-vmware-kubernetes-122-node) are currently in preview phase and we're looking for [feedback](https://github.com/bottlerocket-os/bottlerocket#contact-us).
```
nit: we can drop the preview note here, these are fully supported now

etungsten committed:
QUICKSTART, README: remove notes about vmware variant's preview status

vmware-k8s variants have graduated from preview status.

etungsten commented:
Push above removes note about vmware variant's preview status from README and QUICKSTART-VMWARE.

zmrow reviewed (approved):
:thermometer:
